### PR TITLE
Accumulated fixes for minor issues

### PR DESCRIPTION
**Pull request by JohnoKing**
This commit applies various accumulated bugfixes. Details below:
- Applied some fixes for compiler warnings based off of the following pull requests (with whitespace changes excluded to avoid inflating the size of the diff):
https://github.com/att/ast/pull/281
https://github.com/att/ast/pull/283
https://github.com/att/ast/pull/303
https://github.com/att/ast/pull/304

- `clone_type()`: Two separate variables in this function share the same name. A bugfix from ksh93v- 2013-05-24 was backported to avoid conflict issues.

- Backported a minor error message improvement from ksh2020 for when the user attempts to run too many coprocesses.

- Backported a ksh2020 bugfix for a file descriptor leak: https://github.com/att/ast/commit/58bc8b569d0686439155b3f77e052f773e337367

- Backported ksh2020 bugfixes for unused variables and pointless assignment lint warnings:
https://github.com/att/ast/commit/47650fe0dfa51f25aa90fbb33d734620f705781c
https://github.com/att/ast/commit/df209c0d62af49f236a035a6ee46d73d96abac5d
https://github.com/att/ast/commit/5e417b00bba9ca8e5ce8b245153e916e3cfab81c

- Applied a few minor improvements to libast from graphviz:
https://gitlab.com/graphviz/graphviz/-/commit/e7c0354185cb2af3800ce702804829af1c3e6577
https://gitlab.com/graphviz/graphviz/-/commit/969a7cdeba64342cc4b23852db91b7a0465ae4e9

- `dtmemory()`: Mark unused parameters with `NOT_USED()`. Based on:
https://gitlab.com/graphviz/graphviz/-/commit/6ac3ad99e624d59a786daa3fa7fa0a253f17fff9

- ~libast: Call the shell interpreter with `sh -c --` where applicable. This prevents issues when executing commands whose names start with a dash. Reference:
https://austingroupbugs.net/view.php?id=1440~

- Applied a few documentation/comment tweaks for the NEWS file, `printf -v` and `spawnveg`.

- Added a missing regression test for using the `rm` builtin's `-f` option without additional arguments (this was fixed in ksh93u+ 2012-02-14).

### CODE REVIEW

File: src/lib/libast/comp/system.c [hunk 1]
```diff
@@ -43,13 +43,14 @@
...
-	char*	sh[4];
+	char*	sh[5];
 
 	if (!cmd)
 		return !eaccess(pathshell(), X_OK);
 	sh[0] = "sh";
 	sh[1] = "-c";
-	sh[2] = (char*)cmd;
-	sh[3] = 0;
+	sh[2] = "--";
```
Comment (McDutchie) [inline src/lib/libast/comp/system.c]: This breaks on FreeBSD:

```
$ sh -c -- 'echo hi'
echo hi: --: not found
```

See: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=220587

They have been refusing to fix it for years because their ports scripts are broken. :-/

File: src/cmd/ksh93/bltins/typeset.c [hunk 2]
```diff
@@ -1090,9 +1092,9 @@ int	b_builtin(int argc,char *argv[],Shbltin_t *context)
 	struct tdata tdata;
 	Shbltin_f addr;
 	Stk_t	*stkp;
-	void *library=0;
 	char *errmsg;
 #if SHOPT_DYNAMIC
+	void *library=0;
 	unsigned long ver;
 	int list = 0;
 	char path[1024];
```
Comment (McDutchie) [inline src/cmd/ksh93/bltins/typeset.c]: The `library` and `ver` variables and the `path` array are only used in the `if(arg)` block on lines 1159-1169, so it's probably better to define them within that block. Plus, the `1024` should be `PATH_MAX`.

File: src/lib/libast/tm/tmxfmt.c [hunk 3]
```diff
@@ -137,7 +137,6 @@ tmxfmt(char* buf, size_t len, const char* format, Time_t t)
 	{
 		if ((c = *format++) == delimiter)
 		{
-			delimiter = 0;
```
Comment (McDutchie) [inline src/lib/libast/tm/tmxfmt.c:140]: I'm not comfortable with this deletion. Are you sure it's correct?

Comment (JohnoKing) [reply src/lib/libast/tm/tmxfmt.c:140]: It should be harmless. The `for(;;)` loop it's in is rather gigantic, so it can be hard to follow. The logic behind is pretty simple though. It starts [here](https://github.com/ksh93/ksh/blob/24d5cd853e3661c704abb55ed596d03d51c4e722/src/lib/libast/tm/tmxfmt.c#L135-L138) at L135-138, and when the break at [L140-142](https://github.com/ksh93/ksh/blob/24d5cd853e3661c704abb55ed596d03d51c4e722/src/lib/libast/tm/tmxfmt.c#L140-L142) happens it skips to the code after the for loop, at [L711-715](https://github.com/ksh93/ksh/blob/24d5cd853e3661c704abb55ed596d03d51c4e722/src/lib/libast/tm/tmxfmt.c#L711-L715). If this happens, then `delimiter` is not used after the `break;`. If that condition isn't fulfilled, then the code at [L143-146](https://github.com/ksh93/ksh/blob/24d5cd853e3661c704abb55ed596d03d51c4e722/src/lib/libast/tm/tmxfmt.c#L143-L146) sets `delimiter` to the value of `sp->delimiter`, making the initial `delimiter = 0` redundant.

Comment (McDutchie) [reply src/lib/libast/tm/tmxfmt.c:140]: Thanks. Okay, LGTM.
